### DISCUSSION TITLE
Found a space in the home path - why the warning and how to fix

### DISCUSSION THREAD

**Discussion by Gitznik** [pypa/pipx#1330] [General]
# Intro

In pipx version 1.5, we introduced the warning you're seeing, as multiple incompatibilites with spaces in the pipx home path were discovered. You may see this for the following reasons:

# Why are you seeing this error

1. From pipx version 1.3 to 1.5, we were by default using a path with a space on it on MacOS. This unfortunately means, that all users that installed pipx in this time frame and were using the default behavior are seeing this warning now.
2. You set your `PIPX_HOME` to a path with spaces in it explicitly

# Why are spaces in the `PIPX_HOME` path bad

The main reason we can't support paths with spaces is that shebangs don't support spaces in the interpreter path. All applications installed via `pipx` are installed via `pip`, which creates a script with a shebang at the top, defining the interpreter of the `venv` to use.

`pip` does some magic to the shebang for scripts defined as a `script`, that resolves this issue. Unfortunately, many libraries define their scripts as `console_scripts`, where `pip` does not perform this logic. Therefore, these scripts cannot be run if installed with `pipx` in a path with spaces, as the path to the `venv` and therefore the interpreter to use will contain spaces.

If you want to use a script installed via pipx in a shebang itself (common for example for the aws cli), you run into a similar problem, as the path to the installed script will contain a space.

See these issues for more details on the described problems: https://github.com/pypa/pipx/issues/1198, https://github.com/pypa/pipx/issues/1212

# How to fix

You can generally fix this by using our default locations.

If you're really sure you want to stick to your path with spaces, you can run all `pipx` command with the `-q` flag, ignoring warnings.

> [!WARNING]  
> While the following scripts have been tested and should work reliably, please save the result of `pipx list --json` before running them, so worst case at least you have all the information you need to replicate your old installs.

## MacOS

```bash
cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
logs_dir=$(pipx environment --value PIPX_LOG_DIR)
trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
home_dir=$(pipx environment --value PIPX_HOME)
rm -rf "$cache_dir" "$logs_dir" "$trash_dir"
mkdir -p ~/.local && mv "$home_dir" ~/.local
pipx reinstall-all
```

## Linux

```bash
cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
logs_dir=$(pipx environment --value PIPX_LOG_DIR)
trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
home_dir=$(pipx environment --value PIPX_HOME)
if [ -n "$XDG_DATA_HOME" ]; then
    data_home="$XDG_DATA_HOME"
else
    data_home="$HOME/.local/share"
fi
rm -rf $cache_dir $logs_dir $trash_dir
mkdir -p $data_home && mv $home_dir $data_home
pipx reinstall-all
```

## Windows

Requires bash (for example `git bash`)

> [!WARNING]  
> This has not been tested by a maintainer. Please let us know if this works for you.

```bash
cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
logs_dir=$(pipx environment --value PIPX_LOG_DIR)
trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
home_dir=$(pipx environment --value PIPX_HOME)
rm -rf $cache_dir $logs_dir $trash_dir
mv $home_dir ~
pipx reinstall-all
```

**Comment by watercrossing** [comment]
I have never had a problem with a space in my home directory on windows with powershell. Is there a way to mute the warning?

**Reply by Gitznik** [reply]
Not at the moment. Introducing an environment variable to mute this sounds like a nice idea.

**Reply by Gitznik** [reply]
This will be available with the next release, via the `PIPX_HOME_ALLOW_SPACE` environment variable.

**Reply by rpdelaney** [reply]
@Gitznik Which version should I look for this feature in?

**Reply by Gitznik** [reply]
@rpdelaney this will be in the next release.

**Reply by rpdelaney** [reply]
For anyone else, the current version at the time of writing is `v1.7.0`

**Comment by velipso** [comment]
Your MacOS instructions fail because of spaces in the variables.

You need to add double quotes to the instructions at the bottom:

```bash
cache_dir=$(pipx environment --value PIPX_VENV_CACHEDIR)
logs_dir=$(pipx environment --value PIPX_LOG_DIR)
trash_dir=$(pipx environment --value PIPX_TRASH_DIR)
home_dir=$(pipx environment --value PIPX_HOME)
rm -rf "$cache_dir" "$logs_dir" "$trash_dir"
mkdir -p ~/.local && mv "$home_dir" ~/.local
pipx reinstall-all
```

**Reply by Gitznik** [reply]
Updated the instructions, thanks for pointing it out :+1:

**Reply by andrewluetgers** [reply]
this is what I needed thanks!

**Comment by adehad** [comment]
The edge case I have run into is when the username has a space (Windows), which is due to our company IT configuration. In these cases the suggested commands would still produce issues.

For Windows I ran the `pipx environment` command to get the list of all the paths used, then I ran the following to set all the relevant env vars (for the current user)

Example:
```cmd
setx PIPX_HOME "V:\packages\pipx"
setx PIPX_BIN_DIR "V:\packages\.local\bin"
setx PIPX_MAN_DIR "V:\packages\.local\share\man"
setx PIPX_SHARED_LIBS "V:\packages\pipx\shared"
setx PIPX_LOCAL_VENVS "V:\packages\pipx\venvs"
setx PIPX_LOG_DIR "V:\packages\pipx\logs"
setx PIPX_TRASH_DIR "V:\packages\pipx\trash"
setx PIPX_VENV_CACHEDIR "V:\packages\pipx\.cache"
setx PIPX_STANDALONE_PYTHON_CACHEDIR "V:\packages\pipx\py"
```

If desired a symlink could also be created to the original home directory location. Or the `pipx ensurepaths` can be run again. (`cmd` only for mklink`)
```
mklink /d %USERPROFILE$\.local "V:\packages\pipx\.local\"
```

**Comment by Gitznik** [comment]
This is now part of our official documentation, hence I'm closing this discussion. For any other problems regarding this, please refer to the `Troubleshooting` section of our docs.

**Reply by mrkmcnamee** [reply]
Hi @Gitznik 

As pointed out in the [comment above](https://github.com/pypa/pipx/discussions/1330#discussioncomment-9456553), the $HOME variable on Windows can also contain a space. So the documentation for moving the installation on Windows does not work for some users (including me). Could the instructions be amended to setting a foolproof location on Windows?
https://pipx.pypa.io/stable/installation/#moving-your-pipx-installation

Cheers,
Kevin

**Reply by mrkmcnamee** [reply]
I tried followed the instructions using `C:\pipx` but it failed:
```powershell
PS C:\Git\my-project> pipx --version  
⚠️ Found a space in the pipx home path. We heavily discourage this, due to multiple incompatibilities. Please check our docs for more information on
    this, as well as some pointers on how to migrate to a different home path.
1.7.1
PS C:\Git\my-project> $HOME
C:\Users\Kevin McNamee
PS C:\Git\my-project> $NEW_LOCATION = "C:\pipx"  
PS C:\Git\my-project> $cache_dir = pipx environment --value PIPX_VENV_CACHEDIR
⚠️ Found a space in the pipx home path. We heavily discourage this, due to
    multiple incompatibilities. Please check our docs for more information on
    this, as well as some pointers on how to migrate to a different home path.
PS C:\Git\my-project> $logs_dir = pipx environment --value PIPX_LOG_DIR
⚠️ Found a space in the pipx home path. We heavily discourage this, due to
    multiple incompatibilities. Please check our docs for more information on
    this, as well as some pointers on how to migrate to a different home path.
PS C:\Git\my-project> $trash_dir = pipx environment --value PIPX_TRASH_DIR
⚠️ Found a space in the pipx home path. We heavily discourage this, due to
    multiple incompatibilities. Please check our docs for more information on
    this, as well as some pointers on how to migrate to a different home path.
PS C:\Git\my-project> $home_dir = pipx environment --value PIPX_HOME
⚠️ Found a space in the pipx home path. We heavily discourage this, due to
    multiple incompatibilities. Please check our docs for more information on
    this, as well as some pointers on how to migrate to a different home path.
PS C:\Git\my-project> Remove-Item -Recurse -Force -ErrorAction SilentlyContinue "$cache_dir", "$logs_dir", "$trash_dir"
PS C:\Git\my-project> Remove-Item -Recurse -Force -ErrorAction SilentlyContinue "$NEW_LOCATION"
PS C:\Git\my-project> Move-Item -Path $home_dir -Destination "$NEW_LOCATION"
PS C:\Git\my-project> pipx reinstall-all
⚠️ Found a space in the pipx home path. We heavily discourage this, due to multiple incompatibilities. Please check our docs for more information on
    this, as well as some pointers on how to migrate to a different home path.
No packages reinstalled after running 'pipx reinstall-all' 😴
S C:\Git\my-project> pipx environment
⚠️ Found a space in the pipx home path. We heavily discourage this, due to multiple incompatibilities. Please check our docs for more information on
    this, as well as some pointers on how to migrate to a different home path.
Environment variables (set by user):

PIPX_HOME=
PIPX_GLOBAL_HOME=
PIPX_BIN_DIR=
PIPX_GLOBAL_BIN_DIR=
PIPX_MAN_DIR=
PIPX_GLOBAL_MAN_DIR=
PIPX_SHARED_LIBS=
PIPX_DEFAULT_PYTHON=
PIPX_FETCH_MISSING_PYTHON=
USE_EMOJI=
PIPX_HOME_ALLOW_SPACE=

Derived values (computed by pipx):

PIPX_HOME=C:\Users\Kevin McNamee\pipx
PIPX_BIN_DIR=C:\Users\Kevin McNamee\.local\bin
PIPX_MAN_DIR=C:\Users\Kevin McNamee\.local\share\man
PIPX_SHARED_LIBS=C:\Users\Kevin McNamee\pipx\shared
PIPX_LOCAL_VENVS=C:\Users\Kevin McNamee\pipx\venvs
PIPX_LOG_DIR=C:\Users\Kevin McNamee\pipx\logs
PIPX_TRASH_DIR=C:\Users\Kevin McNamee\pipx\trash
PIPX_VENV_CACHEDIR=C:\Users\Kevin McNamee\pipx\.cache
PIPX_STANDALONE_PYTHON_CACHEDIR=C:\Users\Kevin McNamee\pipx\py
PIPX_DEFAULT_PYTHON=C:\Users\Kevin McNamee\AppData\Local\Microsoft\WindowsApps\PythonSoftwareFoundation.Python.3.11_qbz5n2kfra8p0\python.exe
USE_EMOJI=true
PIPX_HOME_ALLOW_SPACE=false
```

Is PIPX_HOME meant to point to the new location now?
Also, should the old pipx directory be removed as well?

**Reply by mrkmcnamee** [reply]
The following commands are also needed:
```Powershell
> $ENV:PIPX_HOME = "C:\pipx" 
> pipx ensurepath
```
But these settings are not persistent.

**Reply by Gitznik** [reply]
> The following commands are also needed:
> 
> ```Powershell
> 
> > $ENV:PIPX_HOME = "C:\pipx" 
> 
> > pipx ensurepath
> 
> ```
> 
> But these settings are not persistent.

Hey there. You need to set the new `PIPX_HOME` as a persistent environment variable. I'm not sure how to do that on windows, but that should be straight forward to look up. 

Feel free to adjust the docs in a PR once you've figured it out 👍
